bip-0322.mediawiki

```diff
@@ -13,119 +13,99 @@
-# Let sighash = sha256(sha256(scriptPubKey || pre-image))
+* All signatures must use the SIGHASH_ALL flag.
+* The proof is considered valid, inconclusive, or invalid based on whether the to_sign transaction is a valid spend of the to_spend transaction or not, according to the rules specified in the "Consensus and standard flags" section below.
+* It may (optionally) be encumbered with the in_future flag, according to the rules specified in the "Locktime and Sequence" section below, in which case we refer to the result in text form as "valid_in_future", "inconclusive_in_future", etc.
 
-A private key may be used directly to sign a message. In this case, its P2WPKH bech32 address shall be derived, and used as the input.
+Proofs with additional inputs are the base64-encoding of the to_spend and to_sign transactions concatenated in standard network serialisation, and proofs without additional inputs (simple proofs) are the base64-encoding of the to_sign script witness.
 
-=== Signing ===
+A validator must verify it is valid and meets the description of virtual transactions as specified above. See "Validation" below.
 
-The signature is generated as follows:
+=== Validation ===
 
-# Derive the private key privkey for the scriptPubKey; FAIL if not VALID
-# Generate and return a signature sig with privkey=privkey, sighash=sighash
-
-=== Verifying ===
+To validate a proof with additional inputs, the following steps must be taken:
 
-Verify a proof, given a standard flags value, a script sig, an optional witness, and a derived sighash as described above.
+# deserialize the to_spend and to_sign transactions from the proof, and fail if the proof contains extraneous bytes
+# verify that any additional inputs being proven (proof of funds) are included as inputs in the to_sign transaction, exactly once
+# reconstruct the to_spend' and to_sign' transactions, based on the specification above, copying the version, lock time, and sequence values
+# verify that to_spend = to_spend', that to_sign has at least 1 input, has exactly 1 output, and that to_sign.vin[0] = to_sign'.vin[0]
+# optional: set the "in_future" flag if the transaction's lock time is in the future according to consensus rules
+# in "coins", a mapping of outpoints (hash, vout) to coins (scriptPubKey, amount), let coins(to_spend.txid, 0) = (to_spend.vout[0], 0)
+# for each proof of fund input, set the corresponding values in the coins map; abort if the input cannot be found
+# check the signature of each input using consensus rules, then upgradable rules
 
-While omitted below, ERROR is returned if an unforeseen error occurs at any point in the process. A concrete example of this is if a legacy proof is given as input to a non-legacy address; the deserialization of the proof will fail in this case, and this should result in an ERROR result.
+To validate a simple proof, the following steps must be taken:
```

Comment:
I think simple proof have to come first, so if the reader is only interested in basic "simple proof" usecase, they will be presented with it immediately, rather than read a long list of items and then realize "hey, I do not need all that, I only need these two items"

Reply:
Of course, swapped.